README.md
# angular-feedback

[![Bower version](https://badge.fury.io/bo/angular-feedback.svg)](http://badge.fury.io/bo/angular-feedback)
[![Build Status](https://travis-ci.org/andreipfeiffer/angular-feedback.svg?branch=master)](https://travis-ci.org/andreipfeiffer/angular-feedback)
[![Coverage Status](https://coveralls.io/repos/andreipfeiffer/angular-feedback/badge.png?branch=master)](https://coveralls.io/r/andreipfeiffer/angular-feedback?branch=master)

A simple, lightweight module for displaying notifications and loaders in your AngularJS app.

[View Demo](http://andreipfeiffer.github.io/angular-feedback/)

This is a personal edit of the __ng-notify__ module, by __matowens__. For a list of changes, please see the [changes](#changes-to-ng-notify) section at the bottom.

For the original module, please see: [https://github.com/matowens/ng-notify](https://github.com/matowens/ng-notify)

### Requirements

AngularJS is the only dependency.  All animations are CSS-based.

# Installation

You can install using Bower.

    bower install angular-feedback --save

Or, you can download source files straight from this repo, they're located in the `dist` folder.
Just include the minified version of both `.js` and `.css` files.

# Usage

After including **feedback.min.js** and **feedback.min.css**, inject the `feedback` provider into your project.

    // inject it in your application module
    var app = angular.module('myApp', ['feedback']);

    // then inject and use it in your controllers
    app.controller('MainCtrl', ['feedback', '$http', function(feedback, $http) {

        feedback.load();

        $http.get('some/request').success(function () {
            feedback.notify('Your request has completed!');
        });
    }]);

Now you can trigger notifications and loaders from anywhere in your app.

### 1. Notifications

To display a notification, just use the `notify` method.

    feedback.notify('Your notification message goes here!');

To specify the **type** of notification to display, provide the optional **type** param. (For available types, check the [methods](#methods) below.)

    feedback.notify('Your error message goes here!', 'error');

### 2. Loaders

To display a loader, just use the `load` method.

    feedback.load();

To dismiss a loader, or a notification, just use the `dismiss` method.

    feedback.dismiss();

# Advanced Usage

### Set Default Configuration

You can override the default options for all notifications by using the `config` method. None of these options are required. (For available options, check the [config method](#config-options-) below.)

    feedback.config({
        type: 'success',
        duration: 3000,
        sticky: false
    });

### Individual Configurations

You can also pass an object of options to individual notifications. You can pass through any combination of the available options here as well. For example:

    feedback.notify('Your first message.', {
        type: 'warn',
        duration: 1000
    });

### Sticky Notifications

Sticky notifications allow you to set a persistant notification that doesn't fade away. To do this, simply set the `sticky` attribute to true:

    feedback.notify('This is sticky.', {
        sticky: true
    });

This will give the user the option of closing the notification themselves. If you need to dismiss a notification manually, you can do so with the `dismiss` method like this:

    feedback.dismiss();

**NOTE:**
*Any time a notification is set to sticky, the duration attribute will be ignored since the notification will not be automatically fading out.*

# Methods

#### config( options )
Sets default settings for all notifications to take into account when displaying.
- **options** - *object* - an object of options that overrides the default settings.
    - **type**: *string* - *optional* - sets the default notification type when a type isn't explicitly set.
        - info *(default)*
        - error
        - success
        - warn
        - grimace
    - **duration**: *integer* - *optional* - the duration the notification stays visible to the user, in milliseconds.
    - **sticky**: *boolean* - *optional* - determines whether or not the message will fade at the end of the duration or if the message will persist until the user dismisses it themselves. When true, duration will not be set, even if it has a value. *(false by default)*.

#### notify( message, [type] )
Displays a notification message and sets the formatting/behavioral options for this one notification.
- **message**: *string* - *required* - the message to display in your notification.
- **type**: *string* - *optional* - the type of notification to display.
    - info *(default)*
    - error
    - success
    - warn
    - grimace

#### load()
Displays a loader element.

#### dismiss()
Manually removes any notification or loader, by fading them away. The animation is different, depending weather it's a loader or a notification.

# Getter helpers

#### isActive()
Returns `true` if there is any type of feedback displayed (notification of loader). Otherwise, returns `false`.

#### isLoading()
Returns `true` if there is loader displayed. Otherwise, returns `false`.

#### isSticky()
Returns the value of the *sticky* option. Defaults to `undefined` if a loader is active, or no feedback active.

#### getType()
Returns the value of the *type* option. Defaults to `undefined` if a loader is active, or no feedback active.

#### getMessage()
Returns the text message. Defaults to `''`.

# Browser Support

* Firefox 34.*
* Chrome 39.*
* Internet Explorer 10

# Changes to ng-notify

Here is a list of the additions and removals from the original module:

* **Removed the `position` option**. I felt it was something I didn't need anymore, since I have changed the position of the elements. Anyone can override the CSS, to adjust the placement and the styles to their needs.
* **Removed the `theme` option**. Theming the module should be out of scope, from my point of view. The module should not bundle multiple themes. Theming is a presentational concept, so it should be controlled from CSS.
* **Removed dismiss trigger**. I prefer to click the whole notification to dismiss.
* **Move all animations to CSS**. Its easier and more convenient to control animations from CSS.
* **Add "loader" feature**. Usually you display notifications on a callback of a request. And requests often need a loader. So, I felt that these to feature should go together.
* **Add getter methods**. A bunch of public methods were added, to allow devs to query the state and options. Also, very useful for testing.
* **Add unit-tests**. 100% code coverage.


CHANGELOG.md
#angular-feedback changelog

##v0.9.4
- update styling, for Firefox optimization

##v0.9.3
- change module name to: `feedback`

##v0.9.2
- coveralls integration

##v0.9.1
- travis-ci integration

##v0.9.0
- add unit-tests

##v0.8.0
- change name of the package to: `angular-feedback`
- write the README.md file
- fill the CHANGELOG.md file
- register the module as a `bower` package

##v0.7.6
- just refactor & simplify code

##v0.7.5
- **new stuff**: added even more getters: isSticky(), getType(), getMessage()

##v0.7.4
- **new stuff**: added state getters: isActive(), isLoading()

##v0.7.3
- minor styling changes
- add a wrapper to the markup, to allow easier animations

##v0.7.2
- re-write the way animations are performed, to allow action sequencing and ditch the need for $timeout.

##v0.7.1
- Gruntfile updates

##v0.7.0
- change module name to `feedback`
- make notifications re-start animation, when a notification is triggered while other is displayed (using $timeout)

##v0.6.0
- forked the project from `matowens/ng-notify`
- **new feature**: added loader feature
- **new feature**: added CSS-based animations
- cleanup unused code

##v0.5.0
- **new feature**: added a dismiss method to give additional dev control over how notifications display

##v0.4.2
- update for version bumps
- same pathing fix for the package.json file

##v0.4.1
- pathing fix for bower, bower.json file

##v0.4.0
- **new feature**: sticky notifications, dismissable by the user
- standardized comments throughout module
- organized module structure

##v0.3.0
- **new feature**: configurable options for individual notifications
- renamed type option from *defaultType* to just *type* for consistency
- introduced changelog

##v0.2.1
- better organized for bower distribution

##v0.2.0
- initial bower release


spec/feedbackSpec.js
(function() {

    'use strict';

    describe('feedback provider', function() {

        beforeEach(module('feedback'));

        /*
        var feedbackProvider;

        beforeEach(function() {
            // Here we create a fake module just to intercept and store the provider
            // when it's injected, i.e. during the config phase.
            angular
                .module('fakeModule', function() {})
                .config(['feedbackProvider', function(provider) {
                    feedbackProvider = provider;
                }]);

            // name of the provider module
            module('angular-feedback', 'fakeModule');

            // This actually triggers the injection into fakeModule
            inject(function(){});
        });
        */

        it('should be inactive by default', inject(function(feedback) {
            expect( feedback.isActive() ).toBeFalsy();
            expect( feedback.isLoading() ).toBeFalsy();
            expect( feedback.isSticky() ).toBeFalsy();
            expect( feedback.getType() ).toBeUndefined();
        }));

        it('should set loader', inject(function(feedback) {
            feedback.load();
            expect( feedback.isActive() ).toBeTruthy();
            expect( feedback.isLoading() ).toBeTruthy();
        }));

        it('should dismiss loader', inject(function(feedback, $timeout) {
            feedback.load();
            feedback.dismiss();

            $timeout.flush();

            expect( feedback.isActive() ).toBeFalsy();
            expect( feedback.isLoading() ).toBeFalsy();
        }));

        it('should not set a notification with empty message', inject(function(feedback) {
            feedback.notify('');
            expect( feedback.isActive() ).toBeFalsy();
        }));

        it('should be active when notification displayed', inject(function(feedback) {
            var message = 'Notification message';
            feedback.notify( message );
            expect( feedback.isActive() ).toBeTruthy();
            expect( feedback.isLoading() ).toBeFalsy();
        }));

        it('should dismiss notification', inject(function(feedback, $timeout) {
            feedback.notify('Notification');
            feedback.dismiss();

            $timeout.flush();

            expect( feedback.isActive() ).toBeFalsy();
            expect( feedback.isLoading() ).toBeFalsy();
        }));

        it('should set default notification to "info", if no type provided', inject(function(feedback) {
            var message = 'Notification message';
            feedback.notify( message );
            expect( feedback.getType() ).toBe('info');
            expect( feedback.getMessage() ).toBe( message);
        }));

        it('should set default notification to "info", if wrong type provided', inject(function(feedback) {
            var message = 'Notification message';
            feedback.notify( message, 'wrongType' );
            expect( feedback.getType() ).toBe('info');
        }));

        it('should set passed notification type', inject(function(feedback) {
            var message = 'Notification message';
            feedback.notify( message, 'error' );
            expect( feedback.getType() ).toBe('error');
        }));

        it('should set passed notification object', inject(function(feedback) {
            var message = 'Notification message';
            feedback.notify( message, { type: 'success', sticky: true });
            expect( feedback.getType() ).toBe('success');
            expect( feedback.isSticky() ).toBeTruthy();
        }));

        it('should set default config', inject(function(feedback) {
            feedback.config({
                sticky: true,
                type: 'warn'
            });
            feedback.notify('msg');
            expect( feedback.getType() ).toBe('warn');
            expect( feedback.isSticky() ).toBeTruthy();
        }));

        it('should not set default config, if no object is provided', inject(function(feedback) {
            feedback.config();
            feedback.notify('msg');
            expect( feedback.getType() ).toBe('info');
            expect( feedback.isSticky() ).toBeFalsy();
        }));

        it('should add a new type', inject(function(feedback) {
            feedback.addType('custom', 'customClassName');
            feedback.notify('msg', 'custom');
            expect( feedback.getType() ).toBe('custom');
        }));

        it('should not add a new type without name and className', inject(function(feedback) {
            feedback.addType('custom');
            feedback.notify('msg', 'custom');
            expect( feedback.getType() ).toBe('info');

            feedback.addType('', 'customClassName');
            feedback.notify('msg', 'custom');
            expect( feedback.getType() ).toBe('info');
        }));

    });

})();


Gruntfile.js
module.exports = function(grunt) {
    'use strict';

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-cssmin');
    grunt.loadNpmTasks('grunt-processhtml');
    grunt.loadNpmTasks('grunt-contrib-htmlmin');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-karma');
    grunt.loadNpmTasks('grunt-karma-coveralls');

    grunt.initConfig({

        jshint: {
            files: [
                'Gruntfile.js',
                'src/scripts/**/*.js',
                'demo/**/*.js'
            ],
            options: {
                jshintrc: '.jshintrc'
            }
        },

        clean: {
            dist: ['dist']
        },

        cssmin: {
            build: {
                files: {
                    'dist/feedback.min.css': 'src/styles/feedback.css'
                }
            },

            demo: {
                files: {
                    '_gh-pages/demo.min.css': [
                        'src/styles/feedback.css',
                        'demo/demo.css'
                    ]
                }
            }
        },

        processhtml: {
            build: {
                files: {
                    '_gh-pages/index.html': 'demo/index.html'
                }
            }
        },

        htmlmin: {
            options: {
                removeComments: true,
                collapseWhitespace: true
            },
            index: {
                files: {
                    '_gh-pages/index.html': '_gh-pages/index.html'
                }
            }
        },

        uglify: {
            build: {
                files: {
                    'dist/feedback.min.js': 'src/**/*.js'
                }
            },
            demo: {
                files: {
                    '_gh-pages/demo.min.js': [
                        'bower_components/angular/angular.js',
                        'src/scripts/feedback.js',
                        'demo/demo.js'
                    ]
                }
            }
        },

        karma: {
            unit:  { configFile: 'karma.conf.js', keepalive: true }
        },

        coveralls: {
            options: {
                debug: true,
                coverageDir: 'coverage/lcov/',
                dryRun: false,
                force: true,
                recursive: false
            }
        },

        watch: {
            scripts: {
                files: ['src/**/*', 'demo/**/*'],
                tasks: ['build']
            }
        }

    });

    /* Build process...

    - Lint JS
    - Clean old build
    - Process styles
    - Process scripts
    - Process demo */

    grunt.registerTask('build', [
        'jshint',
        'karma',
        'clean',
        'cssmin:build',
        'uglify:build',
        'demo'
    ]);

    grunt.registerTask('demo', [
        'processhtml',
        'htmlmin',
        'cssmin:demo',
        'uglify:demo'
    ]);


    grunt.registerTask('default', ['jshint', 'karma']);
    grunt.registerTask('dev', ['build', 'watch']);

};


demo/demo.js
var app = angular.module('demo', ['feedback']);

app.controller('MainCtrl', ['$scope', 'feedback',
    function($scope, feedback) {
        'use strict';

        // Custom additons...
        /*
        feedback.addType('notice', 'my-notice-type');
        feedback.set('This is my notice type!', 'notice');
        */
        // Demo notifications...

        $scope.notify = function(type) {
            switch(type) {
                case 'success':
                    feedback.notify('You have successfully logged in!', {
                        type: 'success'
                    });
                    break;
                case 'info':
                    feedback.notify('You have a new message in your inbox.', 'info');
                    break;
                case 'warn':
                    feedback.notify('Please login before accessing that part of the site.', 'warn');
                    break;
                case 'error':
                    feedback.notify('The action you are trying to take does not exist.', 'error');
                    break;
                case 'grimace':
                    feedback.notify('An additional notification type to use.', 'grimace');
                    break;
                default:
                    feedback.notify('This is the current default message type. \nIt can be displayed on multiple lines.');
                    break;
            }
        };

        $scope.duration = 4000;
        $scope.durationOptions = [
            { id: 500, value: '500 ms' },
            { id: 1000, value: '1000 ms' },
            { id: 2000, value: '2000 ms' },
            { id: 4000, value: '4000 ms' },
            { id: 8000 , value: '8000 ms'}
        ];

        $scope.defaultType = 'info';
        $scope.defaultOptions = ['info', 'success', 'warn', 'error', 'grimace'];

        $scope.sticky = false;
        $scope.stickyOptions = [false, true];

        $scope.setDefaultType = function() {
            feedback.config({
                type: $scope.defaultType
            });
        };

        $scope.setDefaultDuration = function() {
            feedback.config({
                duration: $scope.duration
            });
        };

        $scope.setDefaultSticky = function() {
            feedback.config({
                sticky: $scope.sticky
            });
        };

        $scope.dismiss = function() {
            feedback.dismiss();
        };

        $scope.loadAndDismiss = function() {
            feedback.load();
            feedback.dismiss();
        };

        $scope.loadAndNotify = function() {
            feedback.load();
            feedback.notify('You have a new message in your inbox.');
        };

        $scope.setLoader = function() {
            feedback.load();
        };

        $scope.isActive = function() {
            console.log( feedback.isActive() );
        };

        $scope.isLoading = function() {
            console.log( feedback.isLoading() );
        };

        $scope.isSticky = function() {
            console.log( feedback.isSticky() );
        };

        $scope.getType = function() {
            console.log( feedback.getType() );
        };

        $scope.getMessage = function() {
            console.log( feedback.getMessage() );
        };
    }
]);


karma.conf.js
// Karma configuration
module.exports = function(config) {

    'use strict';

    config.set({

        // base path, that will be used to resolve files and exclude
        basePath: './',

        // frameworks to use
        frameworks: ['jasmine'],

        // list of files / patterns to load in the browser
        files: [
            // vendors
            'bower_components/angular/angular.min.js',
            // vendor helpers
            'bower_components/angular-mocks/angular-mocks.js',

            // app
            'src/**/*.js',

            // test specs
            'spec/*Spec.js'
        ],

        // list of files to exclude
        exclude: [
        ],

        preprocessors: {
            'src/scripts/*.js': 'coverage',
        },
        coverageReporter: {
            dir : 'coverage/',
            reporters: [
                { type: 'text' },
                { type: 'html', subdir: 'html' },
                { type: 'lcov', subdir: 'lcov' },
            ]
        },

        // test results reporter to use
        // possible values: 'dots', 'progress', 'junit', 'growl', 'coverage'
        reporters: ['dots', 'coverage'],

        // web server port
        port: 9876,

        // enable / disable colors in the output (reporters and logs)
        colors: true,

        // level of logging
        // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        logLevel: config.LOG_INFO,

        // Start these browsers, currently available:
        // - Chrome
        // - ChromeCanary
        // - Firefox
        // - Opera
        // - Safari (only Mac)
        // - PhantomJS
        // - IE (only Windows)
        browsers: ['PhantomJS'],

        // If browser does not capture in given timeout [ms], kill it
        captureTimeout: 10000,

        // Continuous Integration mode
        // if true, it capture browsers, run tests and exit
        singleRun: true,

        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: false

    });
};

src/scripts/ng-notify.js
/**
 * @license
 * http://matowens.github.io/ng-notify
 * (c) 2014 MIT License, matowens.com
 */
(function() {

    'use strict';

     var module = angular.module('feedback', []);

     module.provider('feedback', function() {

        this.$get = ['$document', '$compile', '$rootScope', '$timeout',

            function($document, $compile, $rootScope, $timeout) {

                var options = {
                    duration: 3000,
                    type: 'info',
                    sticky: true
                };

                var userOpts = {};

                var types = {
                    infoClass: 'fdb-info',
                    errorClass: 'fdb-error',
                    successClass: 'fdb-success',
                    warnClass: 'fdb-warn',
                    grimaceClass: 'fdb-grimace',
                    neutralClass: 'fdb-neutral'
                };

                var timeoutAutoDismiss;
                var timeoutDismiss;

                var feedbackScope = $rootScope.$new();
                var tpl = $compile(
                    '<div class="fdb" ng-class="feedbackClass" ng-click="!isLoading && dismiss()">' +
                        '<span class="fdb-message" ng-if="!isLoading">{{ message }}</span>' +
                        '<span class="fdb-spinner" ng-if="isLoading"></span>' +
                    '</div>'
                )(feedbackScope);

                $document.find('body').append(tpl);

                var setType = function(providedType) {
                    var type = (providedType || options.type) + 'Class';
                    return types[type] || types.infoClass;
                };

                var setDuration = function(providedDuration) {
                    var duration = providedDuration || options.duration;
                    return angular.isNumber(duration) ? duration : 3500;
                };

                var setSticky = function(providedSticky) {
                    var sticky = providedSticky || options.sticky;
                    return sticky ? true : false;
                };

                var reset = function() {
                    feedbackScope.feedbackClass = '';
                    feedbackScope.message = '';
                    feedbackScope.isLoading = false;
                };

                feedbackScope.dismiss = function() {
                    if ( feedbackScope.isLoading ) {
                        feedbackScope.feedbackClass += ' fdb-unloading';
                    } else {
                        feedbackScope.feedbackClass += ' fdb-contract';
                    }

                    timeoutDismiss = $timeout(function() {
                        reset();
                    }, 500);
                };


                var el = tpl;

                /**
                 * Our primary object containing all public API methods and allows for all our functionality to be invoked.
                 */
                var feedbackObject = {

                    config: function(params) {
                        params = params || {};
                        angular.extend(options, params);
                    },

                    notify: function(message, userOpt) {

                        if (!message) {
                            return;
                        }

                        $timeout.cancel(timeoutAutoDismiss);
                        $timeout.cancel(timeoutDismiss);

                        if (typeof userOpt === 'object') {
                            userOpts = {
                                type: userOpt.type || undefined,
                                duration: userOpt.duration || undefined,
                                sticky: userOpt.sticky || undefined
                            };
                        } else {
                            userOpts.type = userOpt;
                        }


                        var sticky = setSticky(userOpts.sticky);
                        var duration = setDuration(userOpts.duration);
                        var c = setType(userOpts.type) + ' ';
                        c += sticky ? ' fdb-sticky' : '';
                        c += ' fdb-animate';

                        feedbackScope.feedbackClass = '';
                        feedbackScope.isLoading = false;

                        if (!sticky) {
                            timeoutAutoDismiss = $timeout(function() {
                                feedbackScope.dismiss();
                            }, duration);
                        }

                        $timeout(function() {
                            feedbackScope.feedbackClass = c;
                            feedbackScope.message = message;
                        }, 50);
                    },

                    load: function() {
                        $timeout.cancel(timeoutAutoDismiss);
                        $timeout.cancel(timeoutDismiss);

                        var c = setType('neutral') + ' ' + ' fdb-loading';

                        feedbackScope.feedbackClass = '';
                        feedbackScope.message = '';

                        $timeout(function() {
                            feedbackScope.feedbackClass = c;
                            feedbackScope.isLoading = true;
                        }, 50);
                    },

                    dismiss: function() {
                        feedbackScope.dismiss();
                    },

                    addType: function(typeName, typeClass) {
                        if (!typeName || !typeClass) { return; }
                        types[typeName + 'Class'] = typeClass;
                    }

                };

                return feedbackObject;
            }
        ];
     });
})();


src/scripts/feedback.js
/**
 * @license

 * https://github.com/andreipfeiffer/angular-feedback
 * for current version

 * http://matowens.github.io/ng-notify
 * (c) 2014 MIT License, matowens.com
 * for original version
 */
(function() {

    'use strict';

     var module = angular.module('feedback', []);

     module.provider('feedback', function() {

        this.$get = ['$document', '$compile', '$rootScope', '$timeout',

            function($document, $compile, $rootScope, $timeout) {

                var defaults = {
                    duration: 3000,
                    type: 'info',
                    sticky: false
                };

                var userOpts = {};

                var types = {
                    'info': 'fdb-info',
                    'error': 'fdb-error',
                    'success': 'fdb-success',
                    'warn': 'fdb-warn',
                    'grimace': 'fdb-grimace',
                    'neutral': 'fdb-neutral'
                };

                var timeoutAutoDismiss;
                var timeoutDismiss;

                var feedbackScope = $rootScope.$new();

                feedbackScope.isActive = false;

                var tpl = $compile(
                    '<div class="fdb" ng-show="isActive">' +
                        '<div class="fdb-inner" ng-class="feedbackClass" ng-click="!isLoading && dismiss()">' +
                            '<span class="fdb-message" ng-if="!isLoading">{{ message }}</span>' +
                            '<span class="fdb-spinner" ng-if="isLoading"></span>' +
                        '</div>' +
                    '</div>'
                )(feedbackScope);

                $document.find('body').append(tpl);

                var reset = function() {
                    feedbackScope.isActive = false;
                    feedbackScope.feedbackClass = '';
                    feedbackScope.message = '';
                    feedbackScope.isLoading = false;
                    delete feedbackScope.type;
                    delete feedbackScope.sticky;
                };

                feedbackScope.dismiss = function() {
                    if ( feedbackScope.isLoading ) {
                        feedbackScope.feedbackClass += ' fdb-unloading';
                    } else {
                        feedbackScope.feedbackClass += ' fdb-contract';
                    }

                    timeoutDismiss = $timeout(function() {
                        reset();
                    }, 500);
                };


                var el = tpl;

                /**
                 * Our primary object containing all public API methods and allows for all our functionality to be invoked.
                 */
                var feedbackObject = {

                    config: function(params) {
                        params = params || {};
                        angular.extend(defaults, params);
                    },

                    notify: function(message, userOpt) {
                        var options = {};

                        if (!message) {
                            return;
                        }

                        $timeout.cancel(timeoutAutoDismiss);
                        $timeout.cancel(timeoutDismiss);

                        if (typeof userOpt === 'object') {
                            options = angular.extend( {}, defaults, userOpt );
                        } else {
                            options = angular.extend( {}, defaults );
                            if ( userOpt && types[userOpt] ) {
                                options.type = userOpt;
                            }
                        }

                        var c = types[options.type] + ' ';
                        c += options.sticky ? ' fdb-sticky' : '';
                        c += ' fdb-expand';

                        feedbackScope.isActive = true;
                        feedbackScope.feedbackClass = c;
                        feedbackScope.isLoading = false;
                        feedbackScope.message = message;
                        feedbackScope.type = options.type;
                        feedbackScope.sticky = options.sticky;

                        if ( !options.sticky ) {
                            timeoutAutoDismiss = $timeout(function() {
                                feedbackScope.dismiss();
                            }, options.duration);
                        }
                    },

                    load: function() {
                        $timeout.cancel(timeoutAutoDismiss);
                        $timeout.cancel(timeoutDismiss);

                        var c = types.neutral + ' ' + ' fdb-loading';

                        feedbackScope.isActive = true;
                        feedbackScope.feedbackClass = c;
                        feedbackScope.message = '';
                        feedbackScope.isLoading = true;
                        delete feedbackScope.type;
                        delete feedbackScope.sticky;
                    },

                    dismiss: function() {
                        feedbackScope.dismiss();
                    },

                    isActive: function() {
                        return !!feedbackScope.isActive;
                    },

                    isLoading: function() {
                        return !!feedbackScope.isLoading;
                    },

                    isSticky: function() {
                        return !!feedbackScope.sticky;
                    },

                    getType: function() {
                        return feedbackScope.type;
                    },

                    getMessage: function() {
                        return feedbackScope.message;
                    },

                    addType: function(typeName, typeClass) {
                        if (!typeName || !typeClass) {
                            return;
                        }
                        types[typeName] = typeClass;
                    }

                };

                return feedbackObject;
            }
        ];
     });
})();
